lib/puma/dsl.rb

```diff
@@ -1422,9 +1422,13 @@ def process_hook(options_key, key, block, meth)
 
     def warn_if_in_single_mode(hook_name)
       return if @options[:silence_fork_callback_warning]
-      # user_options (CLI) have precedence over config file
-      workers_val = @config.options.user_options[:workers] || @options[:workers] ||
-        @config.puma_default_options[:workers] || 0
+
+      # TODO: Move this to Configuration and ensure the warnings are only logged
+      #       once the entire configuration is loaded.
+      #
+      #       See https://github.com/puma/puma/issues/3571#issuecomment-2565032734
+      #       for an example of why this isn't a good place to log these warnings.
+      workers_val = @config.instance_variable_get(:@options)[:workers]
```

Comment:
We should get this addressed before merging

Reply:
Done, fortunately there's some good coverage around this.